File: ios/UncompressReactNative.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(UncompressReactNative, NSObject)

RCT_EXTERN_METHOD(
                  extract:(nonnull NSString *)filePath
                  to:(nonnull NSString *)destination
                  overwrite:(BOOL)overwrite
                  password:(NSString *)password
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject
                  )

RCT_EXTERN_METHOD(
                  isProtected:(nonnull NSString *)filePath
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject
                  )

@end

File: ios/UncompressReactNative.swift
@objc(UncompressReactNative)
class UncompressReactNative: NSObject {
    
    // MARK: - Private Properties
    
    private let uncompress = ComponentFactory.uncompressUseCase()
    
    // MARK: - Public Methods

    @objc
    func extract(
        _ filePath: String,
        to destination: String,
        overwrite: Bool = true,
        password: String? = nil,
        resolver resolve: RCTPromiseResolveBlock,
        rejecter reject: RCTPromiseRejectBlock
    ) {
        do {
            try uncompress.extract(
                filePath,
                to: destination,
                overwrite: overwrite,
                password: password
            )
            resolve(NSNull())
        } catch {
            reject("EXTRACT_THROW", error.localizedDescription, error)
        }
    }
    
    @objc
    func isProtected(
        _ filePath: String,
        resolver resolve: RCTPromiseResolveBlock,
        rejecter reject: RCTPromiseRejectBlock
    ) {
        do {
            let isProtected = try uncompress.isProtected(filePath)
            resolve(isProtected)
        } catch {
            reject("IS_PROTECTED_THROW", error.localizedDescription, error)
        }
    }
    
}
